Feature/sfat 238 cloudtrail

adrianmilne opened the pull request:
This is an initial commit candidate for SFAT-238 (but not the whole thing). 

This will create an account specific CloudTrail trail, log them in S3 and push the events to CloudWatch.

I have implemented one alarm - based on S3 activity. There is also an SNS topic (which TechOps can use with OpsGenie). I have tested this by subscribing my email and receive an alert if I drop a file into an S3 bucket.

I thought this might be enough for a tick in the box if it is part of any NFT testing next week.

There is a list of alerts in the JIRA ticket - intention would be to follow up and implement the rest of them (have asked John Denford if they have any examples of Terraform - as they are suggesting these alerts, and that they be terraformed - no reply yet).

Please - any suggestions for improvements/changes gratefully received! 😄

adrianmilne committed:
SFAT-238 - initial CloudTrail configuration

adrianmilne committed:
SFAT-238 - reverted unintended change

tommyb82 commented on ccs-scale-infra-shared/terraform/environments/sbx1/main.tf:
```diff
@@ -31,3 +31,9 @@ module "deploy" {
   environment    = local.environment
 }
+
+module "cloudtrail" {
```
Ahh I was wondering why `terraform plan` on SBX2 was proposing no changes.. I'll copy in now I guess this is intentional (i.e. we only need this auditing on higher environments)?

adrianmilne replied:
I thought so - I assume it's chargeable infra that's not required on the other environments. I did think about building it into another config - which I could look at doing still (that's why they are there I guess) - 'configs/deploy-all-cloudtrail' or something maybe?

tommyb82 commented on ccs-scale-infra-shared/terraform/modules/cloudtrail/main.tf:
```diff
@@ -0,0 +1,174 @@
+#
+# Cloud Trail and CloudWatch Alarms
+#########################################################
+provider "aws" {
```
Was there a reason for re-declaring the provider / switch role here?

adrianmilne replied:
@tommyb82 - I thought it was needed initially - as the current provider is set in the config, and this sidestepped that (originally). However - I wondered if it really was needed if this was called after that code - but I couldn't get it to work - so I have refactored - see comment below

tommyb82 reviewed (approved):
Looks good, provisioned fine, capturing events it seems - just that one comment re. the provider 👍

adrianmilne committed:
SFAT-238 - added CloudTrail to main deply-all, with low defaults

adrianmilne commented:
@tommyb82 - I've refactored the code somewhat. Based on your review comment - I looked into whether I needed the extra `provider` defined - which led to a number of dead ends and unclear issues (I was trying to change as-was and also looked at creating a new config - each had problems). I think these may have been fixed by a complete tear down and rebuild - but then I wondered if it wasn't just better to include CloudTrail into the main deploy anyway, but with 1 day retention defaults. Then it will be consistently tested on each deployment and be available in each environment. 

So this is that refactor - I've set overrides for log retention defaults to 30 days for NFT. If the response from TechOps on the retention for org level trails does not meet NFT requirements - then we can boost this to 7 years (and possibly subsequently add some sensible transitions to Glacier). Should be easy if we need to.

I've also added basic encryption on the S3 bucket (`all data is encrypted at rest`)

tommyb82 commented:
Looks like `count` for modules is coming soon - out in v0.13 public beta at present: https://www.hashicorp.com/blog/announcing-the-terraform-0-13-beta/
Could help with this I guess - set to `0` (zero) for lower environments, 1 for the higher envs?

adrianmilne commented:
> Looks like `count` for modules is coming soon - out in v0.13 public beta at present: https://www.hashicorp.com/blog/announcing-the-terraform-0-13-beta/
> Could help with this I guess - set to `0` (zero) for lower environments, 1 for the higher envs?

Good find - yes, that would certainly. Just tried again to see if it just got ignored to see if we could add it in preparation - but it doesn't take it yet = `The name "count" is reserved for use in a future version of Terraform.`.

What do you think? Keep this approach and create a comment / tech debt note to review after v0.13 is released?

tommyb82 commented:
Yep sounds good - definitely worth revisiting when that feature comes out - could be useful elsewhere too.   👍

adrianmilne committed:
SFAT-238 - added comment for disabling CloudTrail option in TF 0.13.
